File: python/mops/benchmarks/sap.py
import numpy as np
from benchmark import benchmark, format_mean_std

from mops import sparse_accumulation_of_products as sap

# from mops.reference_implementations import sparse_accumulation_of_products as ref_sap


np.random.seed(0xDEADBEEF)

A = np.random.rand(32000, 13)
B = np.random.rand(32000, 7)
C = np.random.rand(900)

indices_A = np.random.randint(13, size=(900,))
indices_B = np.random.randint(7, size=(900,))
output_size = 100
indices_output = np.sort(np.random.randint(output_size, size=(900,)))

# ref_mean, ref_std = benchmark(lambda: ref_sap(A, B, C, indices_A, indices_B, indices_output, output_size))
mean, std = benchmark(lambda: sap(A, B, C, indices_A, indices_B, indices_output, output_size))

# print("Reference implementation:", format_mean_std(ref_mean, ref_std))
print("Optimized implementation:", format_mean_std(mean, std))

# print("Speed-up:", ref_mean / mean)

File: mops/benchmarks/sap.cpp
#include <algorithm>
#include <vector>

#include "mops.hpp"
#include "utils.hpp"

int main() {

    auto A = std::vector<double>(32000 * 13);
    fill_vector_random_floats(A);

    auto B = std::vector<double>(32000 * 7);
    fill_vector_random_floats(B);

    auto C = std::vector<double>(900);
    fill_vector_random_floats(C);

    auto indices_A = std::vector<int32_t>(900);
    fill_vector_random_integers(indices_A, 13);

    auto indices_B = std::vector<int32_t>(900);
    fill_vector_random_integers(indices_B, 7);

    auto indices_output = std::vector<int32_t>(900);
    fill_vector_random_integers(indices_output, 100);

    auto O = std::vector<double>(32000 * 100);

    for (int i = 0; i < 1000; i++) {
        mops::sparse_accumulation_of_products<double>(
            {O.data(), {32000, 100}},
            {A.data(), {32000, 13}},
            {B.data(), {32000, 7}},
            {C.data(), {900}},
            {indices_A.data(), {900}},
            {indices_B.data(), {900}},
            {indices_output.data(), {900}}
        );
    }

    return 0;
}
